File: src/isalgraph/core/sparse_graph.py
"""Adjacency-set sparse graph representation.

Migration of the original ``sparsegraph.py`` with the following bug fix:

* **B1**: ``_edge_count`` was initialized to 1 instead of 0.

Additionally, ``is_isomorphic`` no longer uses ``print()`` for diagnostics —
it returns ``bool`` only.  Callers who need diagnostics should use the
NetworkX adapter's ``nx.is_isomorphic`` for cross-validation.
"""

from __future__ import annotations

import logging

log = logging.getLogger(__name__)


class SparseGraph:
    """High-performance adjacency-set graph with contiguous integer node IDs.

    Optimized for O(1) average edge insertion and membership testing.
    Supports both directed and undirected semantics.

    Args:
        max_nodes: Upper bound on node count (pre-allocates storage).
        directed_graph: Whether edges are directed.
    """

    __slots__ = (
        "_adjacency",
        "_max_nodes",
        "_node_count",
        "_edge_count",
        "_directed_graph",
    )

    def __init__(self, max_nodes: int, directed_graph: bool) -> None:
        self._max_nodes: int = max_nodes
        self._directed_graph: bool = directed_graph
        self._adjacency: list[set[int]] = [set() for _ in range(max_nodes)]
        self._node_count: int = 0
        # BUG FIX B1: was 1 in the original code.
        self._edge_count: int = 0

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------

    def directed(self) -> bool:
        """Return whether the graph is directed."""
        return self._directed_graph

    def node_count(self) -> int:
        """Return the current number of nodes."""
        return self._node_count

    def edge_count(self) -> int:
        """Return the raw stored edge count.

        For undirected graphs each logical edge is counted twice
        (once per direction).  Use ``logical_edge_count()`` for the
        user-facing count.
        """
        return self._edge_count

    def logical_edge_count(self) -> int:
        """Return the number of logical edges.

        For directed graphs this equals ``edge_count()``.
        For undirected graphs this equals ``edge_count() // 2``.
        """
        if self._directed_graph:
            return self._edge_count
        return self._edge_count // 2

    def max_nodes(self) -> int:
        """Return the pre-allocated maximum node capacity."""
        return self._max_nodes

    def neighbors(self, node: int) -> set[int]:
        """Return the adjacency set of *node* (read-only view intended)."""
        if node < 0 or node >= self._node_count:
            raise IndexError(f"Invalid node ID: {node}")
        return self._adjacency[node]

    def has_edge(self, source: int, target: int) -> bool:
        """Return whether the directed edge *source* -> *target* exists."""
        if source < 0 or source >= self._node_count:
            raise IndexError(f"Invalid source node ID: {source}")
        if target < 0 or target >= self._node_count:
            raise IndexError(f"Invalid target node ID: {target}")
        return target in self._adjacency[source]

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def add_node(self) -> int:
        """Add a new node and return its integer ID.

        Raises:
            RuntimeError: If the graph has reached ``max_nodes``.
        """
        if self._node_count >= self._max_nodes:
            raise RuntimeError(f"Maximum number of nodes reached: {self._max_nodes}")
        node_id: int = self._node_count
        self._node_count += 1
        return node_id

    def add_edge(self, source: int, target: int) -> None:
        """Add edge *source* -> *target*.

        For undirected graphs, the reverse edge *target* -> *source* is
        also inserted and the stored edge count is incremented by 2.

        Raises:
            IndexError: If either node ID is out of range.
        """
        if source < 0 or source >= self._node_count:
            raise IndexError(f"Invalid source node ID: {source}")
        if target < 0 or target >= self._node_count:
            raise IndexError(f"Invalid target node ID: {target}")

        if target not in self._adjacency[source]:
            self._adjacency[source].add(target)
            self._edge_count += 1

            if not self._directed_graph:
                self._adjacency[target].add(source)
                self._edge_count += 1

    # ------------------------------------------------------------------
    # Isomorphism (backtracking)
    # ------------------------------------------------------------------

    def is_isomorphic(self, other: SparseGraph) -> bool:
        """Test structural isomorphism with *other* via backtracking.

        This is a simple implementation for testing/debugging.  For
        production use, prefer the NetworkX adapter's ``nx.is_isomorphic``.
        """
        if not isinstance(other, SparseGraph):
            return False
        if self._directed_graph != other._directed_graph:
            return False
        if self._node_count != other._node_count:
            return False

        n = self._node_count
        if n == 0:
            return True

        self_deg = [len(self._adjacency[u]) for u in range(n)]
        other_deg = [len(other._adjacency[u]) for u in range(n)]

        if sorted(self_deg) != sorted(other_deg):
            return False

        # Order by degree descending for early pruning.
        self_order = sorted(range(n), key=lambda u: self_deg[u], reverse=True)
        other_order = sorted(range(n), key=lambda u: other_deg[u], reverse=True)

        mapping: dict[int, int] = {}
        used: set[int] = set()

        def _backtrack(i: int) -> bool:
            if i == n:
                return True
            u = self_order[i]
            for v in other_order:
                if v in used:
                    continue
                if self_deg[u] != other_deg[v]:
                    continue
                # Check consistency with already-mapped neighbours.
                ok = True
                for u2, v2 in mapping.items():
                    if (u2 in self._adjacency[u]) != (v2 in other._adjacency[v]):
                        ok = False
                        break
                    if self._directed_graph and (
                        (u in self._adjacency[u2]) != (v in other._adjacency[v2])
                    ):
                        ok = False
                        break
                if not ok:
                    continue
                mapping[u] = v
                used.add(v)
                if _backtrack(i + 1):
                    return True
                del mapping[u]
                used.remove(v)
            return False

        return _backtrack(0)

    # ------------------------------------------------------------------
    # Dunder
    # ------------------------------------------------------------------

    def __repr__(self) -> str:
        return (
            f"{self.__class__.__name__}("
            f"nodes={self._node_count}, "
            f"edges={self.logical_edge_count()}, "
            f"directed={self._directed_graph})"
        )

File: docs/js/core/sparse-graph.js
(function () {
  'use strict';
  window.IsalGraph = window.IsalGraph || {};

  /**
   * Adjacency-set sparse graph with contiguous integer node IDs.
   *
   * Faithful port of isalgraph.core.sparse_graph.SparseGraph.
   * BUG FIX B1: _edgeCount initialized to 0 (was 1 in original).
   *
   * @param {number} maxNodes - Upper bound on node count.
   * @param {boolean} directedGraph - Whether edges are directed.
   */
  function SparseGraph(maxNodes, directedGraph) {
    this._maxNodes = maxNodes;
    this._directedGraph = directedGraph;
    this._adjacency = [];
    for (var i = 0; i < maxNodes; i++) {
      this._adjacency.push(new Set());
    }
    this._nodeCount = 0;
    // BUG FIX B1: was 1 in original code
    this._edgeCount = 0;
  }

  // ---- Accessors ----

  SparseGraph.prototype.directed = function () {
    return this._directedGraph;
  };

  SparseGraph.prototype.nodeCount = function () {
    return this._nodeCount;
  };

  SparseGraph.prototype.edgeCount = function () {
    return this._edgeCount;
  };

  /**
   * Return the number of logical edges.
   * For undirected graphs: edgeCount / 2.
   */
  SparseGraph.prototype.logicalEdgeCount = function () {
    if (this._directedGraph) {
      return this._edgeCount;
    }
    return this._edgeCount / 2;
  };

  SparseGraph.prototype.maxNodes = function () {
    return this._maxNodes;
  };

  /**
   * Return the adjacency Set for a node.
   * @param {number} node
   * @returns {Set<number>}
   */
  SparseGraph.prototype.neighbors = function (node) {
    if (node < 0 || node >= this._nodeCount) {
      throw new RangeError('Invalid node ID: ' + node);
    }
    return this._adjacency[node];
  };

  /**
   * Check if an edge exists from source to target.
   * @param {number} source
   * @param {number} target
   * @returns {boolean}
   */
  SparseGraph.prototype.hasEdge = function (source, target) {
    if (source < 0 || source >= this._nodeCount) {
      throw new RangeError('Invalid source node ID: ' + source);
    }
    if (target < 0 || target >= this._nodeCount) {
      throw new RangeError('Invalid target node ID: ' + target);
    }
    return this._adjacency[source].has(target);
  };

  // ---- Mutation ----

  /**
   * Add a new node and return its integer ID.
   * @returns {number}
   */
  SparseGraph.prototype.addNode = function () {
    if (this._nodeCount >= this._maxNodes) {
      throw new Error('Maximum number of nodes reached: ' + this._maxNodes);
    }
    var nodeId = this._nodeCount;
    this._nodeCount += 1;
    return nodeId;
  };

  /**
   * Add edge source -> target.
   * For undirected graphs, also adds target -> source (edgeCount += 2).
   * Duplicate edges are silently ignored.
   *
   * @param {number} source
   * @param {number} target
   */
  SparseGraph.prototype.addEdge = function (source, target) {
    if (source < 0 || source >= this._nodeCount) {
      throw new RangeError('Invalid source node ID: ' + source);
    }
    if (target < 0 || target >= this._nodeCount) {
      throw new RangeError('Invalid target node ID: ' + target);
    }

    // Duplicate guard
    if (!this._adjacency[source].has(target)) {
      this._adjacency[source].add(target);
      this._edgeCount += 1;

      if (!this._directedGraph) {
        this._adjacency[target].add(source);
        this._edgeCount += 1;
      }
    }
  };

  // ---- D3-friendly edge list ----

  /**
   * Return edge list as [{source, target}] with no duplicates for undirected.
   * @returns {Array<{source: number, target: number}>}
   */
  SparseGraph.prototype.getEdgeList = function () {
    var edges = [];
    var seen = new Set();
    for (var u = 0; u < this._nodeCount; u++) {
      var iter = this._adjacency[u].values();
      var result = iter.next();
      while (!result.done) {
        var v = result.value;
        if (this._directedGraph) {
          edges.push({ source: u, target: v });
        } else {
          // For undirected, only add each edge once (smaller id first)
          var key = Math.min(u, v) + ',' + Math.max(u, v);
          if (!seen.has(key)) {
            seen.add(key);
            edges.push({ source: Math.min(u, v), target: Math.max(u, v) });
          }
        }
        result = iter.next();
      }
    }
    return edges;
  };

  // ---- Isomorphism (backtracking) ----

  /**
   * Test structural isomorphism with another SparseGraph via backtracking.
   * @param {SparseGraph} other
   * @returns {boolean}
   */
  SparseGraph.prototype.isIsomorphic = function (other) {
    if (!(other instanceof SparseGraph)) return false;
    if (this._directedGraph !== other._directedGraph) return false;
    if (this._nodeCount !== other._nodeCount) return false;

    var n = this._nodeCount;
    if (n === 0) return true;

    var selfDeg = [];
    var otherDeg = [];
    for (var i = 0; i < n; i++) {
      selfDeg.push(this._adjacency[i].size);
      otherDeg.push(other._adjacency[i].size);
    }

    var selfSorted = selfDeg.slice().sort(function (a, b) { return a - b; });
    var otherSorted = otherDeg.slice().sort(function (a, b) { return a - b; });
    for (var j = 0; j < n; j++) {
      if (selfSorted[j] !== otherSorted[j]) return false;
    }

    // Order by degree descending for early pruning
    var selfOrder = [];
    var otherOrder = [];
    for (var k = 0; k < n; k++) {
      selfOrder.push(k);
      otherOrder.push(k);
    }
    var self = this;
    selfOrder.sort(function (a, b) { return selfDeg[b] - selfDeg[a]; });
    otherOrder.sort(function (a, b) { return otherDeg[b] - otherDeg[a]; });

    var mapping = {};
    var used = new Set();

    function backtrack(idx) {
      if (idx === n) return true;
      var u = selfOrder[idx];
      for (var oi = 0; oi < otherOrder.length; oi++) {
        var v = otherOrder[oi];
        if (used.has(v)) continue;
        if (selfDeg[u] !== otherDeg[v]) continue;

        var ok = true;
        for (var u2 in mapping) {
          u2 = parseInt(u2, 10);
          var v2 = mapping[u2];
          if (self._adjacency[u].has(u2) !== other._adjacency[v].has(v2)) {
            ok = false;
            break;
          }
          if (self._directedGraph &&
              (self._adjacency[u2].has(u) !== other._adjacency[v2].has(v))) {
            ok = false;
            break;
          }
        }
        if (!ok) continue;

        mapping[u] = v;
        used.add(v);
        if (backtrack(idx + 1)) return true;
        delete mapping[u];
        used.delete(v);
      }
      return false;
    }

    return backtrack(0);
  };

  // ---- Clone for trace snapshots ----

  /**
   * Deep copy of this graph.
   * @returns {SparseGraph}
   */
  SparseGraph.prototype.clone = function () {
    var copy = new SparseGraph(this._maxNodes, this._directedGraph);
    copy._nodeCount = this._nodeCount;
    copy._edgeCount = this._edgeCount;
    for (var i = 0; i < this._maxNodes; i++) {
      copy._adjacency[i] = new Set(this._adjacency[i]);
    }
    return copy;
  };

  SparseGraph.prototype.toString = function () {
    return 'SparseGraph(nodes=' + this._nodeCount +
           ', edges=' + this.logicalEdgeCount() +
           ', directed=' + this._directedGraph + ')';
  };

  // Export
  IsalGraph.SparseGraph = SparseGraph;
})();
